controllers/atlasmigration_controller.go

```diff
@@ -16,39 +16,310 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
 	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
 
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	dbv1alpha1 "github.com/ariga/atlas-operator/api/v1alpha1"
+	"github.com/ariga/atlas-operator/internal/atlas"
 )
 
+// CLI is the interface used to interact with Atlas CLI
+type MigrateCLI interface {
+	Apply(ctx context.Context, data *atlas.ApplyParams) (*atlas.ApplyReport, error)
+}
+
 // AtlasMigrationReconciler reconciles a AtlasMigration object
 type AtlasMigrationReconciler struct {
 	client.Client
+	CLI    MigrateCLI
 	Scheme *runtime.Scheme
 }
 
+// HCLTmplData is the data used to render the HCL template
+// that will be used for Atlas CLI
+type HCLTmplData struct {
```

Comment:
this one doesn't need to public, and the name should describe the intention. So, I suggest: `atlasMigrationData`. And the template file should be rename to `atlas_migration.tmpl`. That mean, it's template for `atlas.hcl` file for versioned migration.

Reply:
After thinking, if we put the logic for generating atlas.hcl to `internal/atlas`, it becomes clearer and reusable for atlasschema too. What do you think?

Reply:
Yes, but not in this PR. We can rewrite the logic and share for tf provider too